### PR TITLE
*: fix lost index bug of insert on duplicate key update

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs <crazycs520@gmail.com>

### What problem does this PR solve?

Fix issue https://github.com/pingcap/tidb/issues/16669

### What is changed and how it works?

When check the untouched index, should also check the `memory-buffer` in the session too. 

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- No

### CODE REVIEW

File: table/tables/index.go [hunk 1]
```diff
@@ -293,7 +293,11 @@ func (c *index) Create(sctx sessionctx.Context, rm kv.RetrieverMutator, indexedV
 		// If the index kv was untouched(unchanged), and the key/value already exists in mem-buffer,
 		// should not overwrite the key with un-commit flag.
 		// So if the key exists, just do nothing and return.
-		_, err = txn.GetMemBuffer().Get(ctx, key)
+		if memTxn, ok := txn.(kv.MemBufferRetriever); ok {
+			_, err = memTxn.GetFromMemBuffer(ctx, key)
```
Comment (coocood) [inline table/tables/index.go:297]: I think it's better to change TxnState.GetMembuffer implementation, so we will not misuse it later.

File: executor/union_scan_test.go [hunk 2]
```diff
@@ -290,3 +290,31 @@ func (s *testSuite7) TestUnionScanForMemBufferReader(c *C) {
...
+
+	checkFunc := func() {
+		tk.MustExec("begin")
+		tk.MustExec("insert into t values ('a', 1), ('b', 3), ('a', 2) on duplicate key update b = b + 1;")
+		tk.MustExec("commit")
+		tk.MustExec("admin check table t")
+
+		// Test for autocommit
+		tk.MustExec("set autocommit=0")
+		tk.MustExec("insert into t values ('a', 1), ('b', 3), ('a', 2) on duplicate key update b = b + 1;")
```
Comment (tiancaiamao) [inline executor/union_scan_test.go:307]: update a = a + 1?

File: table/tables/index.go (hunk 1 above)
Comment (tiancaiamao) [reply table/tables/index.go:297]: The buffer level is too complex to get things right.

Comment (crazycs520) [reply table/tables/index.go]: great

**Comment by coocood** [comment]
LGTM

**Comment by jackysp** [comment]
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

**Comment by bb7133** [comment]
Data Race case in test:
```
 ==================
[2020-04-22T02:18:01.641Z] WARNING: DATA RACE

[2020-04-22T02:18:01.641Z] Write at 0x0000055cace0 by goroutine 126:

[2020-04-22T02:18:01.641Z]   github.com/pingcap/tidb/executor/seqtest_test.(*testOOMSuite).registerHook()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/log@v0.0.0-20200117041106-d28c14d3b1cd/log.go:110 +0x271

[2020-04-22T02:18:01.641Z]   github.com/pingcap/tidb/executor/seqtest_test.(*testOOMSuite).SetUpSuite()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/seqtest/seq_executor_test.go:1402 +0x50
[2020-04-22T02:18:01.641Z]   runtime.call32()
[2020-04-22T02:18:01.641Z]       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
[2020-04-22T02:18:01.641Z]   reflect.Value.Call()
[2020-04-22T02:18:01.641Z]       /usr/local/go/src/reflect/value.go:321 +0xd3
[2020-04-22T02:18:01.641Z]   github.com/pingcap/check.(*suiteRunner).runFixture.func1()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:799 +0x195
[2020-04-22T02:18:01.641Z]   github.com/pingcap/check.(*suiteRunner).forkCall.func1()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739 +0x113
[2020-04-22T02:18:01.641Z] 
[2020-04-22T02:18:01.641Z] Previous read at 0x0000055cace0 by goroutine 8:
[2020-04-22T02:18:01.641Z]   github.com/pingcap/tidb/domain.(*Domain).topologySyncerKeeper.func1()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/logutil/log.go:344 +0x72
[2020-04-22T02:18:01.641Z]   github.com/pingcap/tidb/domain.(*Domain).topologySyncerKeeper()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/domain/domain.go:488 +0x644
[2020-04-22T02:18:01.641Z] 
[2020-04-22T02:18:01.641Z] Goroutine 126 (running) created at:
[2020-04-22T02:18:01.641Z]   github.com/pingcap/check.(*suiteRunner).forkCall()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:734 +0x4a3

[2020-0422T02:18:01.641Z]   github.com/pingcap/check.(*suiteRunner).runFixture()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:751 +0x83
[2020-04-22T02:18:01.641Z]   github.com/pingcap/check.(*suiteRunner).run()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:695 +0x127
[2020-04-22T02:18:01.641Z]   github.com/pingcap/check.Run()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/run.go:150 +0x5a
[2020-04-22T02:18:01.641Z]   github.com/pingcap/check.RunAll()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/run.go:117 +0x13db
[2020-04-22T02:18:01.641Z]   github.com/pingcap/check.TestingT()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/run.go:99 +0x751
[2020-04-22T02:18:01.641Z]   github.com/pingcap/tidb/executor/seqtest_test.TestT()
[2020-04-22T02:18:01.641Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/seqtest/seq_executor_test.go:73 +0x163
[2020-04-22T02:18:01.641Z]   testing.tRunner()
[2020-04-22T02:18:01.641Z]       /usr/local/go/src/testing/testing.go:909 +0x199
[2020-04-22T02:18:01.641Z] 
[2020-04-22T02:18:01.642Z] Goroutine 8 (finished) created at:
[2020-04-22T02:18:01.642Z]   github.com/pingcap/tidb/domain.(*Domain).Init()
[2020-04-22T02:18:01.642Z]       /home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/domain/domain.go:747 +0x796
[2020-04-22T02:18:01.642Z]   github.com/pingcap/tidb/session.(*domainMap).Get.func1()
```

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_unit_test/detail/tidb_ghpr_unit_test/33522/pipeline

**Pull request merged**